api/v1alpha1/clusterexternaldns_types.go

```diff
@@ -0,0 +1,107 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=cedns
+type ClusterExternalDNS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec   ClusterExternalDNSSpec   `json:"spec,omitempty"`
+	Status ClusterExternalDNSStatus `json:"status,omitempty"`
+}
+
+func (c *ClusterExternalDNS) GetCondition(conditionType string) *metav1.Condition {
+	return meta.FindStatusCondition(c.Status.Conditions, conditionType)
+}
+
+func (c *ClusterExternalDNS) GetConditions() *[]metav1.Condition {
+	return &c.Status.Conditions
+}
+
+func (c *ClusterExternalDNS) GetGeneration() int64 {
+	return c.Generation
+}
+
+func (c *ClusterExternalDNS) SetCondition(condition metav1.Condition) {
+	api.VerifyAndSetCondition(c, condition)
+}
+
+// ClusterExternalDNSSpec allows users to specify desired the state of a cluster-scoped ExternalDNS deployment.
+type ClusterExternalDNSSpec struct {
+	// ResourceName is the name that will be used for the ExternalDNS deployment and related resources. Will default to the name of the ClusterExternalDNS resource if not specified.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^[a-z0-9][-a-z0-9\.]*[a-z0-9]$`
+	// +kubebuilder:validation:Required
+	ResourceName string `json:"resourceName,omitempty"`
+
+	// TenantID is the ID of the Azure tenant where the DNS zones are located.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Format:=uuid
+	// +kubebuilder:validation:Pattern=`[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`
+	TenantID string `json:"tenantID"`
+
+	// DNSZoneResourceIDs is a list of Azure Resource IDs of the DNS zones that the ExternalDNS controller should manage. These must be in the same resource group and be of the same type (public or private). The number of zones is currently capped at 20 but may be expanded in the future.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:MaxItems:=7
+	// +kubebuilder:validation:items:UniqueItems:=true
+	// +kubebuilder:validation:items:MaxLength:=172
+	// +kubebuilder:validation:items:MinLength:=103
+	// +kubebuilder:validation:items:MaxProperties:=1
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[2] == self[0].split('/')[2])",message="all items must have the same subscription ID"
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[4] == self[0].split('/')[4])",message="all items must have the same resource group"
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[7] == self[0].split('/')[7])",message="all items must be of the same resource type"
+	// +listType:=set
+	DNSZoneResourceIDs []string `json:"dnsZoneResourceIDs"`
```

Comment:
need regex validation that it's a resource Id form?

Reply:
validating via kubebuilder:validation:items:pattern doesn't work - it doesn't actually end up appearing on the final resource definition that gets generated. Tried to use XValidation instead but the rule is too complex even with the lowered number of maximum permitted items.